src/execution/execute-afm.ts

```diff
@@ -1,137 +1,251 @@
 // Copyright (C) 2007-2014, GoodData(R) Corporation. All rights reserved.
-import { get, clone } from 'lodash';
 import * as invariant from 'invariant';
 import * as qs from 'qs';
+import { get, range } from 'lodash';
 import { Execution, AFM } from '@gooddata/typings';
+
 import { XhrModule } from '../xhr';
 
-const PAGE_SIZE = 500;
-const DEFAULT_DIMENSION_COUNT = 2;
+export const DEFAULT_LIMIT = 500;
 
-function getDimensionality(execution: AFM.IExecution): number {
-    const dimensions = get(execution, 'execution.resultSpec.dimensions');
-    return dimensions
-        ? dimensions.length
-        : DEFAULT_DIMENSION_COUNT;
+function unwrapExecutionResponse(executionResponseWrapper: Execution.IExecutionResponseWrapper)
+    : Execution.IExecutionResponse {
+    return executionResponseWrapper.executionResponse;
 }
 
-function getLimit(offset: number[]) {
-    return Array(offset.length).fill(PAGE_SIZE);
+function unwrapExecutionResult(executionResultWrapper: Execution.IExecutionResultWrapper)
+    : Execution.IExecutionResult {
+    return executionResultWrapper.executionResult;
 }
 
-// works only for one or two dimensions
-export function mergePageData(result: any, currentPage: any) {
-    const { paging, data, headerItems } = currentPage.executionResult;
-    const rowOffset = paging.offset[0];
-    if (result.executionResult.data[rowOffset]) { // appending columns to existing rows
-        for (let i = 0; i < data.length; i += 1) {
-            const columns = data[i] as any;
+function validateNumOfDimensions(numOfDimensions: number): void {
+    invariant(
+        numOfDimensions === 1 || numOfDimensions === 2,
+        `${numOfDimensions} dimensions are not allowed. Only 1 or 2 dimensions are supported.`
+    );
+}
 
-            result.executionResult.data[i + rowOffset].push(...columns);
-        }
-    } else { // appending new rows
-        result.executionResult.data.push(...data);
+function replaceLimitAndOffsetInUri(oldUri: string, limit: number[], offset: number[]): string {
+    const [uriPart, queryPart] = oldUri.split(/\?(.+)/);
+    const query = {
+        ...qs.parse(queryPart),
+        limit: limit.join(','),
+        offset: offset.join(',')
+    };
+
+    return uriPart + qs.stringify(query, { addQueryPrefix: true });
+}
+
+export function getNextOffset(limit: number[], offset: number[], total: number[]): number[] {
+    const numOfDimensions = total.length;
+    validateNumOfDimensions(numOfDimensions);
```

Comment:
We dont need to validate here. Its enough to do so in the "public methods".

Reply:
removed